src/SAD/ForTheL/Structure.hs

```diff
@@ -42,58 +42,66 @@ import SAD.Data.Text.Decl
 
 -- | The old .ftl file parser.
 forthel :: FTL [ProofText]
-forthel = repeatM $
-  continue <$> (
+forthel = repeatUntil (pure <$> (
     addSectionHeader Signature signatureTags signature'
     <|> addSectionHeader Definition definitionTags definition
     <|> addSectionHeader Axiom axiomTags axiom
     <|> addSectionHeader Theorem theoremTags theorem
-    <|> introduceMacro
-    </> pretypeVariable)
-  <|> bracketExpression
-  <|> (eof >> return abort)
+    <|> (bracketExpression >>= addSynonym)
+    <|> try introduceMacro
+    <|> pretypeVariable))
```

Comment:
I think replacing `</>` is good: it makes the asymmetry with `try` more explicit. We should maybe consider removing `</>` entirely.

Reply:
Honestly I was just confused about how these operators associated, so I used `try` to understand what was going on.